internal/credential/subtype.go

```diff
@@ -22,20 +21,45 @@ func (t SubType) String() string {
 }
 
 // Subtype uses the provided subtype
-func SubtypeFromType(t string) SubType {
+func SubtypeFromType(t string) Subtype {
 	switch {
 	case strings.EqualFold(strings.TrimSpace(t), VaultSubtype.String()):
 		return VaultSubtype
 	}
 	return UnknownSubtype
 }
 
-func SubtypeFromId(id string) SubType {
-	switch {
-	case strings.HasPrefix(strings.TrimSpace(id), vault.CredentialStorePrefix),
-		strings.HasPrefix(strings.TrimSpace(id), vault.CredentialLibraryPrefix),
-		strings.HasPrefix(strings.TrimSpace(id), vault.DynamicCredentialPrefix):
-		return VaultSubtype
+func SubtypeFromId(id string) Subtype {
+	prefix := id[:strings.Index(id, "_")]
+
+	subtypeMu.RLock()
+	subtype, ok := subtypes[prefix]
+	subtypeMu.RUnlock()
+	if !ok {
+		return UnknownSubtype
+	}
+	return subtype
+}
+
+var (
+	subtypeMu sync.RWMutex
+	subtypes  = make(map[string]Subtype)
+)
+
+// Register registers the prefixes for a Subtype. Register panics if the
+// subtype is unknown.
+func Register(subtype Subtype, prefixes ...string) {
+	subtypeMu.Lock()
+	defer subtypeMu.Unlock()
+
+	switch subtype {
+	case VaultSubtype:
+	default:
+		panic("credential.Register: subtype is unknown ")
+	}
+
+	for _, prefix := range prefixes {
+		prefix = strings.TrimSpace(prefix)
+		subtypes[prefix] = subtype
```

Comment:
Do we want to enforce that you cannot have multiple subtypes registered to the same prefix? Or will we leave that to code reviews as with other resource prefixes

Reply:
Unit tests should detect the problem.